lessons/bn/ecto/changesets.md

```diff
@@ -0,0 +1,291 @@
+     [count: 2, validation: :length, kind: :min, type: :string]}
+  ],
+  data: %Friends.Person<>,
+  valid?: false
+>
+```
+
+আপনি হয়তো আশ্চর্য হচ্ছেন, এরর মেসেজে বিদঘুটে `%{count}` দেখে — এটা অন্যান্য ল্যাংগুয়েজে ট্রান্সলেশান সহজ করার জন্যেই বিদ্যমান; আপনি যদি সরাসরি ইউজারকে এরর দেখাতে চান, সেক্ষেত্রে আপনি এটাকে [`traverse_errors/2`](https://hexdocs.pm/ecto/Ecto.Changeset.html#traverse_errors/2) ব্যবহার করে মানুষের পড়ার উপযুক্ত করতে তুলতে পারেন — ডকুমেন্টে উল্লেখিত উদাহরণটি দেখে নিন।
+
+এছাড়াও, `Ecto.Changeset` আরও কিছু বিল্ট-ইন ভ্যালিডেটর নিম্নে দেয়া হলো:
+
++ validate_acceptance/3
++ validate_change/3 & /4
++ validate_confirmation/3
++ validate_exclusion/4 & validate_inclusion/4
++ validate_format/4
++ validate_number/3
++ validate_subset/4
+
+[এখানে](https://hexdocs.pm/ecto/Ecto.Changeset.html#summary), পুরো তালিকাটি দেখতে এবং কিভাবে ব্যবহার করতে হয় তা জানতে পারবেন।
+
+### কাস্টম ভ্যালিডেশান
+
+যদিও বিল্ট-ইন ভ্যালিডেটরগুলো বড় পরিসরের ইউজ কেইস কভার করে, তবুও আপনার হয়তো অন্যরকম কিছু লাগতে পারে।
+
+এতক্ষণ পর্যন্ত আমাদের ব্যবহার করা প্রতিটা `validate_` ফাংশনই `%Ecto.Changeset{}` গ্রহণ ও রিটার্ন করে, তাই আমরা সহজেই আমাদের নিজেদেরটি প্লাগ করতে পারি।
+
+উদাহরণস্বরূপ, শুধুমাত্র ফিকশনাল ক্যারেক্টারের নামই গ্রহণ করা হবে তা আমরা নিশ্চিত করতে পারি:
+
+```elixir
+@fictional_names ["Black Panther", "Wonder Woman", "Spiderman"]
+def validate_fictional_name(changeset) do
+  name = get_field(changeset, :name)
+
+  if name in @fictional_names do
+    changeset
+  else
+    add_error(changeset, :name, "is not a superhero")
+  end
+end
+```
+
+উপরে, আমরা দুইটি নতুন হেল্পার ফাংশন ব্যবহার করেছি: [`get_field/3`](https://hexdocs.pm/ecto/Ecto.Changeset.html#get_field/3) এবং [`add_error/4`](https://hexdocs.pm/ecto/Ecto.Changeset.html#add_error/4)। নাম দেখেই হয়তো এদের কাজ অনুমান করতে পেরেছেন তবুও, আমি আপনাকে ডকুমেন্টেশানের লিঙ্কগুলোতে ঢুঁ মেরে আসার অনুরোধ করবো।
+
+সবসময় `%Ecto.Changeset{}` রিটার্ন করাই উত্তম, তাহলে আপনি `|>` অপারেটর ব্যবহার করতে পারবেন, সুতরাং নতুন নতুন ভ্যালিডেশান যোগ করতে সুবিধা হবে।
+
+```elixir
+def changeset(struct, params) do
+  struct
+  |> cast(params, [:name, :age])
+  |> validate_required([:name])
+  |> validate_length(:name, min: 2)
+  |> validate_fictional_name()
+end
+```
+
+```elixir
+iex> Friends.Person.changeset(%Friends.Person{}, %{"name" => "Bob"})
+%Ecto.Changeset<
+  action: nil,
+  changes: %{name: "Bob"},
+  errors: [name: {"is not a superhero", []}],
+  data: %Friends.Person<>,
+  valid?: false
+>
+```
+
+দারুণ, এটা কাজ করেছে! কিন্তু, আমাদের এই ফাংশন নিজেদের ইমপ্লিমেন্ট করার প্রয়োজন ছিলো না — আমরা চাইলেই `validate_inclusion/4` ব্যবহার করতে পারতাম; যাই হোক, আপনি দেখতে পাচ্ছেন কিভাবে আপনি নিজের এরর দেখাতে পারেন এটা পরবর্তীতে কাজে দিবে।
+
+## প্রোগ্রামেটিকালি চেইঞ্জ যোগ 
+
+অনেক সময়, আপনি হয়তো চেইঞ্জসেটে ম্যানুয়ালি চেইঞ্জ যোগ করতে চাইবেন। এর জন্যে `put_change/3` হেল্পার ফাংশনটি ব্যবহার করতে পারেন।
+
+`name` ফিল্ডটিকে রিকুয়েরড না করে, চলুন ইউজারকে নাম ছাড়াই সাইন আপ করার সুযোগ দিই, এবং তাদের "Anonymous" নামে ডাকি।
```

Comment:
"রিকুয়েরড" নাকি "রিকয়ার্ড"? let's discuss and add to glossary bhai.

Reply:
রিকয়ার্ড sounds better.